opencog/atomspace/atom_types.script

```diff
@@ -76,6 +76,9 @@ EXISTS_LINK <- BIND_LINK
 AVERAGE_LINK <- FORALL_LINK
 SATISFYING_SET_LINK <- BIND_LINK
 
+// Function constructor (predicate and schema)
+LAMBDA_LINK <- BIND_LINK
```

Comment:
OK, there is a fair amount of confusion going on here. You keep thinking and talking in  terms of functional programming and functional evaluation, whereas I am thinking of and talking about satisfiability and satisfaction. These two notions are very different from each other, and this is at the root of the confusion.

Satisfiability is about taking an expression that contains variables, and discovering values for those variables, that renders the expression "true".  So perhaps BindLink should be called "SatisfiabilityLink". The "meaning" of a BindLink is that it encloses an expression that may or may not be true, depending on whether there exist items in the universe that, when substituted for the variables, render the expression true.  Note that there might be multiple values that render a BindLink true: these form a set. Thus, perhaps BindLink should be called "SatisfiabilitySetLink" .  So, in this sense, a BindLink is quite similar to a SatisfyingSetLink.  The main difference between the two is that the BindLink requires that an ImplicationLink be inside of it.  Perhaps this is a design mistake.  Thus, today, all bindLinks must have the following form:

```
BindLink
    VariableDecls x
    ImplicationLink
          P(x)
          Q(x)
```

I'm guessing that what this means, in terms of "classical" PLN, is this:
```
 ForAllLink
     VariableDecls x
     ImplicationLink
         SatisfyingSetLink
               P(x)
          Q(x)
```
Note the indentation of P w.r.t. Q. Basically, if some values satisfy P(x), i.e. if some values belong to the sattisfying set of P(x), then it is true that those values imply Q(x).  The (cog-evaluate!) function does two things: it finds all values v satisfying P(x), (i.e. it finds all values v such that P(v) is true) and then it creates Q(v)  where Q(v) means Q(substitute v for x).

Note in particular, the below is wrong:
```
BindLink
    Variables
    Expression
```
The above is NOT a valid BindLink. the Expression MUST be an ImplicationLink for it to be valid.